### PR TITLE
Fix unsoundness in incremental ALIA

### PR DESCRIPTION
**Pull request by blishko**
Incremental solving including arrays and arithmetic contained a bug that could cause the solver to respond `sat` when the formula is actually `unsat`.
The theory combination relies on creating interface lemmas in order for the individual solvers to agree if a pair of interface variables have the same value or not.
Such interface lemmas are created on demand, lazily. However, the array solver may create new equalities (and introduce new variables to the SAT solver) when it formulates its new lemmas. In case of incremental solving, these outlive their frame in assertion stack after pop. They are still present in the core. If, afterwards, we assert the corresponding disequality (which gets split into inequalities), we miss the introduction of the interface lemmas, because the variables of the equality are not considered interface variables (they appear only in UF part of the formula in first frame, and, after pop, they appear only in the arithmetic part of the formula).

The proposed fix is to introduce interface lemmas in the case of equalities introduced by the array solver more eagerly. This, however, comes with a performance cost because we eagerly add (possibly unnecessary) interface lemmas.

### CODE REVIEW

**Comment by Tomaqa** [comment]
It seems to fix all remaining SMT-LIB benchmarks with arrays (`QF_A*`) under the timeout of 20 minutes. I also have noticed no significant performance drop.

File: src/tsolvers/UFLATHandler.cc [hunk 1]
```diff
@@ -81,34 +74,28 @@ vec<PTRef> UFLATHandler::getSplitClauses() {
...
         return res;
     }
-    if (equalitiesToPropagate.size() == 0) {
-        return res;
-    }
     for (PTRef eq : equalitiesToPropagate) {
-        if (knownEqualities.find(eq) == knownEqualities.end()) {
+        if (equalitiesWithAddedInterfaceClauses.find(eq) == equalitiesWithAddedInterfaceClauses.end()) {
             addInterfaceClausesForEquality(logic, eq, res);
+            equalitiesWithAddedInterfaceClauses.insert(eq);
```
Comment (Tomaqa) [inline src/tsolvers/UFLATHandler.cc:98]:
```suggestion
        if (auto [_, inserted] = equalitiesWithAddedInterfaceClauses.insert(eq); inserted) {
            addInterfaceClausesForEquality(logic, eq, res);
```

File: src/tsolvers/UFLATHandler.cc [hunk 2]
```diff
@@ -81,34 +74,28 @@ vec<PTRef> UFLATHandler::getSplitClauses() {
...
         res.copyTo(arrayLemmas);
-        std::unordered_set<PTRef, PTRefHash> equalitiesWithAddedClauses;
         // HACK: If array solver needs to decide equality on interface vars, we need to add corresponding lemmas already here
         for (PTRef lemma : arrayLemmas) {
             if (logic.isOr(lemma)) {
                 for (PTRef lit : logic.getPterm(lemma)) {
                     PTRef atom = logic.isNot(lit) ? logic.getPterm(lit)[0] : lit;
-                    if (logic.isNumEq(atom) and knownEqualities.find(atom) == knownEqualities.end() and
-                        equalitiesWithAddedClauses.find(atom) == equalitiesWithAddedClauses.end()) {
+                    if (logic.isNumEq(atom) and equalitiesWithAddedInterfaceClauses.find(atom) == equalitiesWithAddedInterfaceClauses.end()) {
```
Comment (Tomaqa) [inline src/tsolvers/UFLATHandler.cc:82]:
```suggestion
                    if (logic.isNumEq(atom) and equalitiesWithAddedInterfaceClauses.count(atom) == 0) {
```

File: src/tsolvers/UFLATHandler.cc [hunk 3]
```diff
@@ -81,34 +74,28 @@ vec<PTRef> UFLATHandler::getSplitClauses() {
...
                         PTRef lhs = logic.getPterm(atom)[0];
                         PTRef rhs = logic.getPterm(atom)[1];
-                        if (std::find(interfaceVars.begin(), interfaceVars.end(), lhs) != interfaceVars.end() and
-                            std::find(interfaceVars.begin(), interfaceVars.end(), rhs) != interfaceVars.end()
-                        ) {
+                        if ((logic.isVar(lhs) or logic.isNumConst(lhs)) and (logic.isVar(rhs) or logic.isNumConst(rhs))) {
```
Comment (Tomaqa) [inline src/tsolvers/UFLATHandler.cc:85]:
```suggestion
                        if (isInterfaceVar(lhs) and isInterfaceVar(rhs)) {
```
using lambda
```
    auto isInterfaceVar = [this](PTRef arg){
        return logic.isVar(arg) or logic.isNumConst(arg);
    };
```
(or use another name)

File: src/tsolvers/egraph/EgraphSolver.cc [hunk 4]
```diff
@@ -1637,12 +1637,7 @@ vec<PTRef> Egraph::collectEqualitiesFor(vec<PTRef> const & vars, std::unordered_
         for (int i = 0; i < equivalentVars.size(); ++i) {
             for (int j = i + 1; j < equivalentVars.size(); ++j) {
                 PTRef eq = logic.mkEq(ERefToTerm(equivalentVars[i]), ERefToTerm(equivalentVars[j]));
-                if (not enode_store.has(eq)) {
-                    if (knownEqualities.find(eq) != knownEqualities.end()) {
-                        throw OsmtInternalException("Internal error in computing interface equalities in Egraph");
-                    }
-                    equalities.push(eq);
-                }
+                if (knownEqualities.find(eq) == knownEqualities.end()) { equalities.push(eq); }
```
[the remark this answers is not in the document]
Comment (blishko) [reply src/tsolvers/egraph/EgraphSolver.cc:1640]: `contains` would be best. Until then, I would prefer `find`. It expresses the intent better (and `count` uses `find` internally at least in one standard library implementation: https://stackoverflow.com/questions/14159682/unordered-map-which-one-is-faster-find-or-count).

File: src/tsolvers/UFLATHandler.cc (hunk 2 above)
Comment (blishko) [reply src/tsolvers/UFLATHandler.cc:82]: See my comment above.

File: src/tsolvers/egraph/EgraphSolver.cc [hunk 5]
```diff
@@ -1637,12 +1637,7 @@ vec<PTRef> Egraph::collectEqualitiesFor(vec<PTRef> const & vars, std::unordered_
         for (int i = 0; i < equivalentVars.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (blishko) [reply src/tsolvers/egraph/EgraphSolver.cc:1636]: Changed!

File: src/tsolvers/UFLATHandler.cc (hunk 1 above)
Comment (blishko) [reply src/tsolvers/UFLATHandler.cc]: Changed!

File: src/tsolvers/UFLATHandler.cc (hunk 3 above)
Comment (blishko) [reply src/tsolvers/UFLATHandler.cc]: Changed!

File: src/tsolvers/egraph/EgraphSolver.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (Tomaqa) [reply src/tsolvers/egraph/EgraphSolver.cc:1639]: I see. OK then.

**Comment by Tomaqa** [comment]
Resolves #710. I have discovered no new bugs even with 1h timeout.

**Pull request merged**